### PR TITLE
Feature/image tags

### LINKED ISSUE #6: Image Tags
#### As a user I want to add tags to images.

__Story__:

Our users are loving the ability to add images to our service. However, now
that there are more than a handful of images, it is apparent that we need some
way to filter images by certain categories. In order to do that, we first need
to label images in some way. It is time to add the ability for our users to tag
images when they add them.

__Acceptance criteria__:
- [x] When I add a new image, there is a form field to add tags.
- [x] After I've saved an image with tags, I see those tags displayed with the
  associated image.
- [x] My application uses the acts-as-taggable-on gem to provide the tagging
  functionality.

__Reference material__:
- https://github.com/mbleigh/acts-as-taggable-on

__Note__: You should allow images to have no tags; this sets the stage for a later issue.

__Dependencies__:
- Image Index

### PR DESCRIPTION
**Pull request by jesse-smidt**
Closes #6 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: app/views/images/show.html.erb [hunk 1]
```diff
@@ -1,3 +1,14 @@
 <h1 class="mt-3">Image Created!</h1>
 <p>An image url has been saved. Access here:</p>
-<%= link_to nil, @image.url %>
+<div>
```
Comment (stevenboyd) [inline app/views/images/show.html.erb]: Right now you are only showing the tags on the show page, not the index page. So you should add them there.

Comment (jesse-smidt) [reply app/views/images/show.html.erb]: Makes sense, but wasn't on the check list. I'll add them in.

Comment (stevenboyd) [reply app/views/images/show.html.erb]: I think one thing to consider, you currently have no way to get back to the show page once you leave. So the only time a user can see the tags is when the first created the image. This isn't really a problem as long as you show the tags also on the index page. But I believe the story assumed the index page was still accessible, possibly by clicking on the image in the index view or something.

File: test/models/image_test.rb [hunk 2]
```diff
@@ -54,4 +54,22 @@ class ImageTest < ActiveSupport::TestCase
     image = Image.new(url: 'http.www.appfolio.com')
     refute image.valid?
   end
+
+  test 'should save with tags' do
```
[the remark this answers is not in the document]
Comment (jesse-smidt) [reply test/models/image_test.rb]: Added check for tag element on index and show actions.

File: test/controllers/application_controller_test.rb [hunk 3]
```diff
@@ -7,4 +7,13 @@ class ApplicationControllerTest < ActionDispatch::IntegrationTest
     assert_response :success
     assert_select 'h1', text: 'Image Sharer'
   end
+
+  test 'should get home with image with tags' do
+    get '/'
```
Comment (stevenboyd) [inline test/controllers/application_controller_test.rb]: Nothing about the setup for this test is different than the `should get home with body` test, so I would combined them.

I would also do 
```ruby
assert_select '.card .badge', text: 'the tag value'
```
instead to also test that the value of the tag is the same.

File: test/controllers/images_controller_test.rb [hunk 4]
```diff
@@ -10,11 +10,13 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
 
   test 'should create new image and respond with redirect' do
     post images_path,
-         params: { image: { url: URL_PATH } }
+         params: { image: { url: URL_PATH, tag_list: 'nice, fun' } }
     assert_response :redirect
     follow_redirect!
     assert_response :success
+    puts response.body
```
Comment (stevenboyd) [inline test/controllers/images_controller_test.rb]: You forgot a `puts`

File: test/controllers/images_controller_test.rb [hunk 5]
```diff
@@ -10,11 +10,13 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
     assert_select 'a', URL_PATH
+    assert_select 'span.badge'
```
Comment (stevenboyd) [inline test/controllers/images_controller_test.rb]: You submitted two tags. I would assert that there are two badges and also assert the text is correct.

File: app/views/application/home.html.erb [hunk 6]
```diff
@@ -11,8 +11,17 @@
 </div>
 <div class="d-flex flex-wrap justify-content-around pt-5">
   <% @images.each do |image| %>
-    <div class="align-self-center m-4 shadow">
+    <div class="align-self-center m-4 shadow card">
       <img class="home__image" src="<%= image.url %>" alt="Image">
+      <% if image.tag_list.any? %>
```
Comment (stevenboyd) [inline app/views/application/home.html.erb]: This is almost identical to the code in the show view. I would push this down into a partial that is shared in these two places.

https://guides.rubyonrails.org/layouts_and_rendering.html#using-partials

File: test/controllers/images_controller_test.rb (hunk 4 above)
Comment (jesse-smidt) [reply test/controllers/images_controller_test.rb]: oops! Removed

File: test/controllers/images_controller_test.rb (hunk 5 above)
Comment (jesse-smidt) [reply test/controllers/images_controller_test.rb]: Updated

File: app/views/application/home.html.erb (hunk 6 above)
Comment (jesse-smidt) [reply app/views/application/home.html.erb]: Updated

File: app/views/partials/_tag_list.html.erb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<% tag_list.each do |tag| %>
+      <span class="badge badge-pill badge-secondary">
```
Comment (stevenboyd) [inline app/views/partials/_tag_list.html.erb]: Minor, but your intending is off here

Comment (jesse-smidt) [reply app/views/partials/_tag_list.html.erb]: Updated

**Pull request merged**